base/functors.jl

```diff
@@ -0,0 +1,135 @@
+###### Functors ######
+
+# Note that functors are merely used as internal machinery to enhance code
+# reuse and improve performance of map/reduce.
+# They are not exported.
+# When function arguments can be inlined, the use of functors can be removed.
+
+abstract Func{N}
+
+immutable IdFun <: Func{1} end
+call(::IdFun, x) = x
+
+immutable AbsFun <: Func{1} end
+call(::AbsFun, x) = abs(x)
+
+immutable Abs2Fun <: Func{1} end
+call(::Abs2Fun, x) = abs2(x)
+
+immutable ExpFun <: Func{1} end
+call(::ExpFun, x) = exp(x)
+
+immutable LogFun <: Func{1} end
+call(::LogFun, x) = log(x)
+
+immutable AndFun <: Func{2} end
+call(::AndFun, x, y) = x & y
+
+immutable OrFun <: Func{2} end
+call(::OrFun, x, y) = x | y
+
+immutable AddFun <: Func{2} end
+call(::AddFun, x, y) = x + y
+
+immutable MulFun <: Func{2} end
+call(::MulFun, x, y) = x * y
+
+immutable MaxFun <: Func{2} end
+call(::MaxFun, x, y) = scalarmax(x,y)
+
+immutable MinFun <: Func{2} end
+call(::MinFun, x, y) = scalarmin(x, y)
+
+# a fallback unspecialized functor that allows code using functors to not care
+# whether they were able to specialize on the function value or not
+immutable UnspecializedFun{N,T<:Callable} <: Func{N}
+    f::T
+end
+call{N,T}(::Type{UnspecializedFun{N}}, f::T) = UnspecializedFun{N,T}(f)
```

Comment:
This causes performance problems with `map(Int, BitArray)` because `specialized_bitwise_unary` returns an `UnspecializedFun{N,DataType}` instead of an `UnspecializedFun{N,Type{Int}}`. There's an obvious fix to make it specialize properly for DataTypes, but I wasn't sure why this type was necessary. Why can't `specialized_unary` etc. just return `f` if there's no corresponding specialized functor?